### PR TITLE
Add a listurls management command

### PR DESCRIPTION
**Pull request by offbyone**
# Trac ticket number

ticket-28800

# Branch description

This picks up work started by @hippietilley on their [hippietilley/django#ticket_28800](https://github.com/hippietilley/django/tree/ticket_28800) branch.

I have made a couple of changes:

* added documentation for the command
* support for multiple URL prefixes
* support for a table output

# Checklist
- [X] This PR targets the `main` branch. 
- [X] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [ ] I have added or updated relevant tests.
- [X] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

# Notes

I was unable to find where tests for management commands belong; happy to add them, but wasn't aware of how.

I'm not sure what documentation other than the manual should be updated; I'd love it if someone pointed me at that.

### CODE REVIEW

**Comment by adamchainz** [comment]
I will try to do a full review during the week. But for now, I would note that you need to squash your commits into one and follow the [commit style guide](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines), and update the PR title to match.

**Comment by offbyone** [comment]
Hopefully this meets approval. I went back and forth on the commit message a few times, apologies for the chaos.

File: docs/releases/5.1.txt [hunk 1]
```diff
@@ -234,6 +234,9 @@ Management Commands
   each operation to highlight :class:`operation categories
   <django.db.migrations.operations.base.OperationCategory>`.
 
+* Introduce a :djadmin:`listurls` command that lists the URLs in the application,
```
Comment (smithdc1) [inline docs/releases/5.1.txt:237]: 5.1 is now in alpha so new features added now will be released in 5.2. Therefore please could you move this to `5.2.txt`?

File: docs/ref/django-admin.txt [hunk 2]
```diff
@@ -498,6 +498,18 @@ Only support for PostgreSQL is implemented.
 
 If this option is provided, models are also created for database views.
 
+``listurls``
+------------
+
+.. django-admin:: listurls [url_prefix ...]
+
```
Comment (smithdc1) [inline docs/ref/django-admin.txt:505]: Add a `versionadded` note. 

```suggestion
------------

.. versionadded:: 5.2

```

File: django/core/management/commands/listurls.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    help = """Displays a list of urls used in the project."""
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "url_prefix",
+            nargs="*",
+            help="Only list URLs with these URL prefixes.",
+        )
+
+        parser.add_argument(
```
Comment (smithdc1) [inline django/core/management/commands/listurls.py:30]: Do we need to have multiple output formats? I couldn't see any discussion on this, so apologies if I've missed something here. 

The table option seems closer to what django-extensions `show_urls` does?

**Review by smithdc1 — COMMENTED** [review]
Thanks for pushing this forward! 🥇 

I've left a few small comments inline. One thing I noticed when testing is that the output with the `--table` option seems a bit surprising. 

In my test project with django-extensions I see. 

```
/       bootstrap4.views.index
/bootstrap3     bootstrap3.views.index  bootstrap3.views.index
```

With the new `listurls` command:

```
        <function index at 0x000002216F5DF600>.index
django  <function index at 0x000002216F634E00>.index    django_rendering.views.index
```

I can see two differences. The urls do not have a leading `/` which I'm not sure is an issue but means that nothing is shown for the `/` case. Also the `view_methods` are incorrectly formatted.

The leading `/` is also an issue for the longer output format where the `URL` field is blank. 

```
URL:
View: bootstrap4.views.index
Arguments:
```

File: django/core/management/commands/listurls.py (hunk 3 above)
Comment (offbyone) [reply django/core/management/commands/listurls.py:30]: The table option is what `show_urls` did, but the long form is what @hippietilley put in their original patch; given I was carrying on their work, it seemed polite to include their format. I'm happy to reverse the defaults, or elide it, but I didn't want to be rude about the work they already did.

**Comment by offbyone** [comment]
I'll look at the output again; I suspect I missed something in the view lookup code and URL formatting. Updates to come.

**Comment by ulgens** [comment]
@offbyone Hey, do you plan to continue with this PR? If not, I'd like to take the issue.

**Comment by offbyone** [comment]
I plan to, but I've been too busy to pick it up. If you want to take what I've got and finish it, I'd be more than happy to hand it over.

**Comment by felixxm** [comment]
Superseded by #18769.

**Pull request closed without merging**